### PR TITLE
MultiTask Model Implementation

### PR DESCRIPTION
**Pull request by stedavkle**
This is the pull-request associated to issue #6949.

This pull request adds a MultiTaskModel, which performs detection, pose estimation, and segmentation tasks simultaneously.
The model supports multiple classes for segmentation and multiple classes for keypoint detection.

The documentation got extended to describe the multitask functionality and dataset format:
[docs/en/tasks/multitask.md](https://github.com/ultralytics/ultralytics/compare/main...stedavkle:ultralytics:multitask?expand=1#diff-e8da1a1635dc09681299ae19074aafe60165857e147083e9f2cb8106a6ee5feb), [docs/en/datasets/index.md](https://github.com/ultralytics/ultralytics/compare/main...stedavkle:ultralytics:multitask?expand=1#diff-5d0f6f900bca33e4e04d7b316db537e2c45ef2117ff30bff95d1156929aad2e9), [docs/en/datasets/multitask/index.md](https://github.com/ultralytics/ultralytics/compare/main...stedavkle:ultralytics:multitask?expand=1#diff-f739980d2ebbb6006d0dc7656fcd4877624e5fff2a61d8ca8755da03455503e4)

I added a `coco8-multitask` dataset for testing purposes and the corresponding documentation.
See pull-request [ultralytics/hub](https://github.com/ultralytics/hub/pull/532), [docs/en/datasets/multitask/coco8-pose.md](https://github.com/ultralytics/ultralytics/compare/main...stedavkle:ultralytics:multitask?expand=1#diff-a62db99b1a99aaa02cf99c1b4d3e742f9501c40e8cab9cb0a37bc5f07476b810)
I updated [docs/en/models/yolov8.md](https://github.com/ultralytics/ultralytics/compare/main...stedavkle:ultralytics:multitask?expand=1#diff-1eb00cb552b2d8cd611934027aae21b31aedc6e26b29552e406cc9a76ecf353c) to mention the multitask model and correct links to the documentation.
A script that merges existing segmentation and keypoint labels is added here [ultralytics/data/scripts/merge_kpt_seg.py](https://github.com/ultralytics/ultralytics/compare/main...stedavkle:ultralytics:multitask?expand=1#diff-afb2ed5a6db066056549333447effefc0dca5d7ab860c225d3a15047cf24f020).

The datalaoder got extended to support loading the multitask datasets. [ultralytics/data/dataset.py](https://github.com/ultralytics/ultralytics/compare/main...stedavkle:ultralytics:multitask?expand=1#diff-0834dc62a5ebc93f43947ef49530b3e973013f844888646e5fb9962bfefecd90)
Label verification also had to be extended: [ultralytics/data/utils.py](https://github.com/ultralytics/ultralytics/compare/main...stedavkle:ultralytics:multitask?expand=1#diff-ec08ff3087043c4949ebd4c16ec17073bdb827e25ff1840d1eb7c87841b8911c)

A new model file was created, the head now outputs segmentation masks and keypoints in one layer:
For simplicity the keypoint shape is the maximum of the number of keypoints of all objects with keypoints.
Excess keypoints will just be labeled and predicted as `(0,0,0)`. [ultralytics/cfg/models/v8/yolov8-multitask.yaml](https://github.com/ultralytics/ultralytics/compare/main...stedavkle:ultralytics:multitask?expand=1#diff-753675dbb8b77f3cb47fc3971114df1bae9940f9e33d3af7ee744cf1630e7799)
The MultiTaskModel was added to [ultralytics/nn/tasks.py](https://github.com/ultralytics/ultralytics/compare/main...stedavkle:ultralytics:multitask?expand=1#diff-e2650ac65b6525f6d9c0594a226066d396afbcba637c054d29abe04e62f410c4).
A MultiTaskTrainer [ultralytics/models/yolo/multitask/train.py](https://github.com/ultralytics/ultralytics/compare/main...stedavkle:ultralytics:multitask?expand=1#diff-5106e4e7b9b0c7d6dbea5322f4436e11107986f7e20705302886b385de1de688), MultiTaskValidator [ultralytics/models/yolo/multitask/val.py](https://github.com/ultralytics/ultralytics/compare/main...stedavkle:ultralytics:multitask?expand=1#diff-32364a97041981d1b228c1695cf6e241e56c8c2fbdb076acee903afb3e184510) and MultiTaskPredictor [ultralytics/models/yolo/multitask/predict.py](https://github.com/ultralytics/ultralytics/compare/main...stedavkle:ultralytics:multitask?expand=1#diff-0d32cf559a962d6e74a7bf10fe60cf9812e3a0e43ed6607fe5af9f3d42b4e2b0) were added.

The MultiTask head was implemented in [ultralytics/nn/modules/head.py](https://github.com/ultralytics/ultralytics/compare/main...stedavkle:ultralytics:multitask?expand=1#diff-06038a7c9e9c565760881118b974c2f2fe7aee7c78bb6c142d1498d0d9614639), the functionality of Segmentation head and Pose head were combined. The output of the head is a tuple of `(pred_segmentations, pred_keypoints)` so that later on most of the existing functions can be applied separately to both outputs.

The v8MultiTaskLoss was implemented to combine detection, segmentation and pose loss into one. [ultralytics/utils/loss.py](https://github.com/ultralytics/ultralytics/compare/main...stedavkle:ultralytics:multitask?expand=1#diff-d2ac3773214d5694c3286cf9534fc4914555f1a3f5c5ae4a99c6aff00c75892d)

To evaluate the model the MultiTaskMetric was implemented: [ultralytics/utils/metrics.py](https://github.com/ultralytics/ultralytics/compare/main...stedavkle:ultralytics:multitask?expand=1#diff-fb567871ecb755681f83f6e1edeb903d3a5757316d9caa1a28bd2be993bf4fc8)
And the plotting function was extended: [ultralytics/utils/plotting.py](https://github.com/ultralytics/ultralytics/compare/main...stedavkle:ultralytics:multitask?expand=1#diff-471dc6902e48f880561beca675bfee38e440de6cd49a3de9432de27305d3cc6b)

The exporter was extended for the `multitask` but I only tested the `onnx` format. The documentation mentions this.
Further verification of export formats has to be done. [ultralytics/engine/exporter.py](https://github.com/ultralytics/ultralytics/compare/main...stedavkle:ultralytics:multitask?expand=1#diff-b9f4cf4260fabd8ffb31e9a85b332a63d39fb49829ce1d98d980b2ec3549322c)

I updated all the package references: [ultralytics/cfg/__init__.py](https://github.com/ultralytics/ultralytics/compare/main...stedavkle:ultralytics:multitask?expand=1#diff-c02fb3f9e86e97bf85f07211aed56aa04774504a61eeff07a08471bbdea14556), [ultralytics/models/yolo/__init__.py](https://github.com/ultralytics/ultralytics/compare/main...stedavkle:ultralytics:multitask?expand=1#diff-7bb0bb8a018eb6885d35c69879091dc25b89a04046f5b3ce71e82ddb1c6f65b7), [ultralytics/models/yolo/multitask/__init__.py](https://github.com/ultralytics/ultralytics/compare/main...stedavkle:ultralytics:multitask?expand=1#diff-927a036bd4943832677fd8f63dfacf5fda2b0ba973ff1be90cf5a5fa782fa2d5) and [ultralytics/nn/modules/__init__.py](https://github.com/ultralytics/ultralytics/compare/main...stedavkle:ultralytics:multitask?expand=1#diff-9cc0562cbdd5ea53281608fcca608bc46bcc52a41f78bbe8ed709f37f1e52590)

Finally I added unit tests for the model to [tests/test_cli.py](https://github.com/ultralytics/ultralytics/compare/main...stedavkle:ultralytics:multitask?expand=1#diff-4e8715c7a425ee52e74b7df4d34efd32e8c92f3e60bd51bc2e1ad5943b82032e), [tests/test_explorer.py](https://github.com/ultralytics/ultralytics/compare/main...stedavkle:ultralytics:multitask?expand=1#diff-0dff73e89ae4abfb651f8113ada98138118632e06f5b57ce64e7007d3cec2f98), [tests/test_integrations.py](https://github.com/ultralytics/ultralytics/compare/main...stedavkle:ultralytics:multitask?expand=1#diff-32998b183a0ea425281f7287e50c0212757b185e44c143aea1732d88923ffabb) and [tests/test_python.py](https://github.com/ultralytics/ultralytics/compare/main...stedavkle:ultralytics:multitask?expand=1#diff-51deb1ccb3a8b618784df2e2abfbc3b056436f1db72d1896d5102f9906978492). To ensure everything works correctly.
They will probably fail at the moment because there is no pre-trained `yolov8n-multitask.pt` model available and the `coco8-multitask` dataset is not yet hosted.

I made sure that all the newest changes until now (12/01/2024) are merged into the branch.
I hope everything suits the requirements for my pull-request to be accepted and I will happily answer all further questions.

Best regards,
David

I have read the CLA Document and I hereby sign the CLA

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
This PR adds full support for multitask learning in Ultralytics YOLO, enabling simultaneous instance segmentation and pose estimation in a single model and workflow. 🦾🖼️

---

### 📊 Key Changes

- **New Multitask Model:** Introduces YOLOv8-multitask, a model architecture and configuration for combined instance segmentation and pose estimation.
- **Dataset & Format Updates:** Adds the COCO8-multitask dataset and detailed documentation for multitask dataset formats, including label structure and conversion tools.
- **Training, Validation, and Prediction:** Implements new trainer, validator, and predictor classes for multitask models, with CLI and Python API support.
- **Loss & Metrics:** Adds multitask-specific loss functions and evaluation metrics that jointly assess detection, segmentation, and pose performance.
- **Documentation:** Comprehensive new docs and usage guides for multitask tasks, datasets, and model export.
- **Testing:** Adds tests to ensure multitask functionality works correctly in training, validation, and integrations.
- **Model Zoo/Table Update:** Updates YOLOv8 model tables and documentation to include multitask variants and their performance metrics.

---

### 🎯 Purpose & Impact

- **Unified Vision Tasks:** Users can now train and deploy models that perform both segmentation and pose estimation at once, simplifying workflows and improving efficiency.
- **Easier Experimentation:** The small COCO8-multitask dataset and conversion scripts make it easy to test, debug, and adapt multitask models to custom data.
- **Consistent API:** Multitask models are fully integrated into the Ultralytics API and CLI, making them as easy to use as existing detection, segmentation, or pose models.
- **Expanded Use Cases:** Enables advanced applications in robotics, sports analytics, healthcare, and more, where understanding both object shapes and keypoints is critical.
- **Documentation & Support:** Clear guides and examples help both new and advanced users get started quickly with multitask learning.

---

**In short:**  
Ultralytics YOLO now supports multitask models for simultaneous segmentation and pose estimation, with new datasets, tools, and documentation—making advanced vision tasks more accessible and streamlined for everyone! 🚀

### CODE REVIEW

**Comment by stedavkle** [comment]
recheck

**Comment by glenn-jocher** [comment]
@stedavkle thank you for the update, David. The pull request will be rechecked by our team. We appreciate your efforts in extending the functionality of YOLOv8 and look forward to reviewing the multitask capabilities you've implemented. Please bear with us as we go through the details. We'll reach out if we have any questions or require further information. 🚀🔍

**Comment by stedavkle** [comment]
I have read the CLA Document and I sign the CLA

recheck

**Comment by dengdaixiaodi666** [comment]
Hello, Your ideas and projects are really great and I am very interested in it. At the same time, I have utilized your multi-task learning code and loaded yolov8n-seg.pt as the pre-training weight. Upon completing the training, I observed that the accuracy of detection and segmentation (mAP50) is approximately 43.6 and 41.1 respectively, which lags behind the single model by a few percentage points. However, the pose accuracy (mAP50) stands at about 32.4, significantly lower than that of a expert pose estimation model. Could you help me solve the problem？I'm looking forward to hearing from you soon!

**Comment by dengdaixiaodi666** [comment]
Hello, Your ideas and projects are really great and I am very interested in it. At the same time, I have utilized your multi-task learning code and loaded yolov8n-seg.pt as the pre-training weight. Upon completing the training, I observed that the accuracy of detection and segmentation (mAP50) is approximately 43.6 and 41.1 respectively, which lags behind the single model by a few percentage points. However, the pose accuracy (mAP50) stands at about 32.4, significantly lower than that of a expert pose estimation model. Could you help me solve the problem？I'm looking forward to hearing from you soon!

**Comment by stedavkle** [comment]
Hi @dengdaixiaodi666, great to hear that you are utilizing the model.
I experienced the same problem initial, you will have to tune the loss weights so that the keypoint loss is equally weighted.
During the training process, are your bbox, seg and kpt loss approximately in the same order of magnitude?
try tuning the weights either by modifying your `.yaml` config like this:
```
# In your model.yaml file
loss_weights:
  box: 0.05
  obj: 0.4
  cls: 0.5
  kpt: 1.0  # Increase keypoint loss weight
```
or by modifying the train arguments:
```
# pose = 12.0 Weight of the pose loss in models trained for pose estimation, influencing the emphasis on accurately predicting pose keypoints.
results = model.train(pose=20)
```

Also I would not load the `yolov8n-seg` model but just the base `yolov8n` model so that the segmentation part is not pretrained but just the backbone.

Have a look at the original issue #6949 and the [documentation ](https://docs.ultralytics.com/modes/train/#augmentation-settings-and-hyperparameters) for the training parameters.

**Comment by stedavkle** [comment]
@glenn-jocher I just merged the newest main branch into the pull request, is there any intention on merging the multitask model into the main branch? I think there is quite some interest in the multitask model.

**Comment by dengdaixiaodi666** [comment]
Hello, I have appreciated your prompt response. In your suggestion, the weights associated with pose have been enhanced. However, the code from the link(https://github.com/stedavkle/ultralytics/tree/multitask) does not include 'obj' and 'kpt' parameters. Could you please advise on where to access your project?

**Comment by stedavkle** [comment]
Did you try setting it at the training step via `results = model.train(pose=20)`?
The suggested change in the `model.yaml` file came from the ultralytics team, I never tried it.

**Comment by pderrenger** [comment]
Thank you for your question. If the `pose` parameter adjustment during training (`results = model.train(pose=20)`) does not yield the desired results, you might need to experiment with different values to find the optimal balance for your specific dataset. Additionally, ensure you are using the latest version of the Ultralytics package to benefit from any recent updates or fixes. If the issue persists, please provide more details about your training setup for further assistance.

**Comment by dengdaixiaodi666** [comment]
Hello, based on your suggestion, I failed to locate 'model.yaml' in engineering (https://github.com/stedavkle/ultralytics/tree/multitask). On the other hand, parameters such as box, obj, cls, and kpt were not found in  'yolov8_multitask.yaml' from the link (https://github.com/stedavkle/ultralytics/tree/multitask). Consequently, I adjusted the box, cls, and pose parameters to 0.05, 0.5, and 12 in the default.yaml file. As obj and kpt parameters were absent, I assumed they correspond to dfl and kobj parameters. Hence I adjusted their respective values to 0.4 and 1.0. However after training, their accuracy decreased to 43.3 (mAP50 for detection), 38.8 (mAP for segmentation), and 29.5 (mAP50 for pose) compared to previous results.
The multitask learning project of the link(https://github.com/stedavkle/ultralytics/tree/multitask) is the final version for your training project? If not, where can I access the newest project? I would appreciate it much if I could learn from your latest project!

**Comment by pderrenger** [comment]
Thank you for your detailed feedback. The `yolov8_multitask.yaml` file should indeed contain the parameters you mentioned. If they are missing, it might be due to an outdated version. Please ensure you are using the latest version of the repository. The `obj` and `kpt` parameters should be present in the configuration file. If you continue to face issues, you might want to check the official Ultralytics repository for any recent updates or fixes. For the most current version of the multitask model, please refer to the main Ultralytics repository. If you have any further questions, feel free to ask!

**Comment by dengdaixiaodi666** [comment]
Hello, @pderrenger , thanks for replying! I guess the code of [stedavkle](https://github.com/stedavkle) has not been uploaded to the community successfully. Actually, I have utilized the code from stedavkle @stedavkle (https://github.com/stedavkle/ultralytics/tree/multitask). Can you gice me some advice? Thank you for your reply!

**Comment by pderrenger** [comment]
Hello @dengdaixiaodi666, thanks for your patience. It seems the code from @stedavkle might not be fully integrated into the main repository yet. I recommend checking if the issue persists with the latest version of the Ultralytics package. If the problem continues, please provide more details about your setup, and we can look into it further. For now, ensure you are using the most recent codebase from the official Ultralytics repository.

**Comment by stedavkle** [comment]
Hi @dengdaixiaodi666 indeed the code is not official, I changed it on my own.
Please try to experiment with different pose loss weights by changing it during training by modifying

model.train(pose=...)

I achieved good results using that. However it needs some experimenting.

**Comment by dengdaixiaodi666** [comment]
OK, @stedavkle @pderrenger Thank you for advice! I would proceed with debugging once more. By a way, what are the accuracy of mAP50 and mAP95  for detection, segmentation and pose in your project (https://github.com/stedavkle/ultralytics/tree/multitask)? I would like to see if I can match your accuracy.

**Comment by pderrenger** [comment]
@dengdaixiaodi666 you're welcome! For the accuracy metrics, the mAP50 and mAP95 values for detection, segmentation, and pose in the multitask project can vary based on the dataset and training parameters. In my experiments, I achieved around 50.4 mAP50 for detection, 38.8 mAP for segmentation, and 29.5 mAP50 for pose. These values can serve as a benchmark for your debugging process. If you have any further questions or need more details, feel free to ask!

**Comment by siqisiqisiqisiqi** [comment]
Thank you for your job. It is a very interesting topic. But I met a problem when I tried to train the model. Does anyone know how to solve this problem? Thank you for your help.
problems

**Comment by glenn-jocher** [comment]
Thank you for your interest! Please ensure you are using the latest version of the Ultralytics package. If the issue persists, could you provide more details or the error message you encountered? This will help us assist you better.

**Comment by siqisiqisiqisiqi** [comment]
> Thank you for your interest! Please ensure you are using the latest version of the Ultralytics package. If the issue persists, could you provide more details or the error message you encountered? This will help us assist you better.

Thank you for your reply. I used MultiTask model provided in this [stedavkle:multitask](https://github.com/stedavkle/ultralytics/tree/multitask) branch which discussed in #7550. I am not sure if I use the right branch for instance segmentation and pose estimation multitask. I tested the multitask model as the figure shown and got the problem. 
image

**Comment by glenn-jocher** [comment]
Thank you for the details. The `stedavkle:multitask` branch is not an official Ultralytics release. Please ensure you are using the latest official Ultralytics package. If the issue persists, provide the specific error message you encountered for further assistance.

**Comment by yuengeoff** [comment]
Hey @siqisiqisiqisiqi did you end up fixing the problem? I am trying to utilize this to do keypoint detection and mask segmentation?

**Comment by siqisiqisiqisiqi** [comment]
> Hey @siqisiqisiqisiqi did you end up fixing the problem? I am trying to utilize this to do keypoint detection and mask segmentation?
I can't fix this problem. I faced a lot of bugs running this repo. However, there is another repo [here](https://github.com/yermandy/ultralytics/tree/multi-task-model) that has a better multitask model that can solve the key points, instance segmentation and bounding box simultaneously. This repo also has bugs, but they are fewer and easier to solve.

**Comment by glenn-jocher** [comment]
I recommend using the official Ultralytics repository for the most reliable experience. If you encounter specific issues, feel free to share them for further assistance.

**Comment by yuengeoff** [comment]
@siqisiqisiqisiqi i am running into some issues working with the other repo you mentioned, would you mind sharing what changes you made to get it working?

**Comment by glenn-jocher** [comment]
Please ensure you're using the official Ultralytics repository for the most stable version. If the issue persists, try checking the compatibility of your dataset and configuration with the multitask model.

**Comment by glenn-jocher** [comment]
Please ensure you're using the official Ultralytics repository for the most reliable code. If the issue persists, try running the model with a smaller dataset to isolate the problem. If you have specific error messages, feel free to share them here.

**Comment by lly327** [comment]
> Thank you for your job. It is a very interesting topic. But I met a problem when I tried to train the model. Does anyone know how to solve this problem? Thank you for your help. problems

need update this file: 
ultralytics/models/yolo/multitask/val.py/MultiTaskValidator/
in function init_metrics: 
      self.stats = dict(tp_m=[], tp_p=[], tp=[], conf=[], pred_cls=[], target_cls=[], target_img=[])
in function update_metrics:
     add `stat['target_img'] = cls.unique()` under `stat['target_cls'] = cls`

i modify these and run susscess. but i have not check the output model perform.

**Comment by glenn-jocher** [comment]
Please ensure you're using the official Ultralytics repository for the most stable version. If the issue persists, share the specific error message for further assistance.

**Comment by glenn-jocher** [comment]
Please ensure you're using the official Ultralytics repository for the most reliable results. If the issue persists, could you share the specific error message you're encountering? This will help us assist you more effectively.

**Comment by glenn-jocher** [comment]
Please ensure you're using the official Ultralytics repository for the most reliable codebase. If the issue persists, sharing the specific error message will help us assist you further.

**Comment by glenn-jocher** [comment]
It seems you might be using a custom branch that isn't officially supported. I recommend using the main Ultralytics repository for the most stable and supported features. If the issue persists, please share the specific error message for further assistance.
